### PR TITLE
Significantly improve error messages for expr

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

**Comment by tpoterba** [comment]
Depends on #2657

File: python/hail/expr/expression.py [hunk 1]
```diff
@@ -3546,77 +3577,82 @@ def analyze(expr, expected_indices, aggregation_axes=set()):
...
-                list(expected_axes),
-                list(indices.axes)
+            "scope violation: '{caller}' expects an expression indexed by {expected}"
+            "\n    Found indices {axes}, with unexpected indices {stray}. Invalid fields:{fields}{agg}".format(
+                caller=caller,
+                expected=list(expected_axes),
+                axes=list(indices.axes),
+                stray=list(unexpected_axes),
+                fields=''.join("\n        '{}' (indices {})".format(name, list(inds.axes)) for name, inds in bad_refs),
+                agg='' if (unexpected_axes - aggregation_axes) else
```
Comment (jbloom22) [inline python/hail/expr/expression.py]: why use this condition for determining whether to print the "supports aggregation over" message?

File: python/hail/api2/matrixtable.py [hunk 2]
```diff
@@ -446,7 +443,7 @@ def annotate_globals(self, **named_exprs):
         base, cleanup = self._process_joins(*named_exprs.values())
 
         for k, v in named_exprs.items():
-            analyze(v, self._global_indices)
+            analyze('MatrixTable.anntoate_globals', v, self._global_indices)
```
Comment (jbloom22) [inline python/hail/api2/matrixtable.py]: typo: anntoate

File: python/hail/api2/matrixtable.py [hunk 3]
```diff
@@ -734,15 +731,15 @@ def select_cols(self, *exprs, **named_exprs):
 
         for e in exprs:
             all_exprs.append(e)
-            analyze(e, self._col_indices, {self._row_axis})
+            analyze('MatrixTable.select_cols', e, self._col_indices, {self._row_axis})
```
Comment (jbloom22) [inline python/hail/api2/matrixtable.py]: you might as well move `def select_rows` above `def select_cols` in this PR so it parallels all the others

File: python/hail/api2/matrixtable.py [hunk 4]
```diff
@@ -1777,12 +1774,13 @@ def joiner(left):
                 raise NotImplementedError('vds join with indices {}'.format(indices))
 
             return construct_expr(Select(Reference(prefix), uid),
-                                  self.row_schema, indices, aggregations, joins.push(Join(joiner, uids_to_delete)))
+                                  self.row_schema, indices, aggregations,
```
Comment (jbloom22) [inline python/hail/api2/matrixtable.py]: just making sure you don't mean to use construct_reference here and below

File: python/hail/expr/aggregators.py [hunk 5]
```diff
@@ -12,18 +12,19 @@ def _to_agg(x):
...
 
 
 @typecheck(name=strlike, aggregable=Aggregable, ret_type=Type, args=anytype)
 def _agg_func(name, aggregable, ret_type, *args):
     args = [to_expr(a) for a in args]
-    indices, aggregations, joins = unify_all(aggregable, *args)
+    indices, aggregations, joins, refs = unify_all(aggregable, *args)
     if aggregations:
-        raise ValueError('cannot aggregate an already-aggregated expression')
+        raise ExpressionException('cannot aggregate an already-aggregated expression')
```
Comment (jbloom22) [inline python/hail/expr/aggregators.py]: For consistency, capitalize Cannot here and in the rest

File: python/hail/api2/matrixtable.py (hunk 4 above)
Comment (tpoterba) [reply python/hail/api2/matrixtable.py]: nope, references are meant to be top-level fields used and this is a UID

File: python/hail/expr/expression.py (hunk 1 above)
Comment (tpoterba) [reply python/hail/expr/expression.py]: This condition means that all stray axes are available for aggregation. It's possible you use GT in a `select_global` or something that doesn't support aggregation. In that case we don't want the message.

**Pull request merged**